### PR TITLE
Allow skipped task state task_instance_schema.py

### LINKED ISSUE #31420: allow other states than success/failed in tasks by REST API
### Apache Airflow version

main (development)

### What happened

see also: https://github.com/apache/airflow/issues/25463

From the conversation there, it sounds like it's intended to be possible to set a task to "skipped" via REST API, but it's not.
Instead the next best thing we have is marking as success & adding a note.

### What you think should happen instead

I see no reason for users to not just be able to set tasks to "skipped". I could imagine reasons to avoid "queued", and some other more "internal" states, but skipped makes perfect sense to me for 
- tasks that failed and got fixed externally
- tasks that failed but are now irrelevant because of a newer run
in case you still want to be able to see those in the future (actually, a custom state would be even better)

### How to reproduce

```python
# task is just a dictionary with the right values for below
r = requests.patch(
    f"{base_url}/api/v1/dags/{task['dag_id']}/dagRuns/{task['dag_run_id']}/taskInstances/{task['task_id']}",
    json={
        "new_state": "skipped",
    },
    headers={"Authorization": token},
)
```
-> r.json() gives
`{'detail': "'skipped' is not one of ['success', 'failed'] - 'new_state'", 'status': 400, 'title': 'Bad Request', 'type': 'http://apache-airflow-docs.s3-website.eu-central-1.amazonaws.com/docs/apache-airflow/latest/stable-rest-api-ref.html#section/Errors/BadRequest'}`

### Operating System

/

### Versions of Apache Airflow Providers

/

### Deployment

Astronomer

### Deployment details

/

### Anything else

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by JoranDox**
closes: #31420

this PR is mostly to accompany my issue #31420

### CODE REVIEW

**Review by hussein-awala — CHANGES_REQUESTED** [review]
Allowing all the TI states is not a good idea, where this can lead to unexpected result.
IMO, the task could be stucked if we change its state manually to `deferred`, `removed`, `shutdown`, ...
You can just add skipped as accepted value in these methods. Don't forget to add a unit test for your change

**Comment by JoranDox** [comment]
Fair enough, I'm just a sucker for generic solutions, but you're right, we probably wouldn't want states like those to be set manually.
I've been looking through the tests but can't seem to find the ones for that part of the api, I feel like they should be in here: https://github.com/apache/airflow/blob/main/airflow/api/common/mark_tasks.py but I think I'm missing something.

File: airflow/api_connexion/schemas/task_instance_schema.py
Comment (ephraimbuddy) [inline airflow/api_connexion/schemas/task_instance_schema.py]: You also need to update the state in the openapi specification https://github.com/apache/airflow/blob/985f3501806980e6612ae7fbe81669b001e93597/airflow/api_connexion/openapi/v1.yaml#L4244-L4249

Comment (JoranDox) [reply airflow/api_connexion/schemas/task_instance_schema.py]: Ah good point, I saw that first before I got to the part I already edited above, but I guess I assumed this was automatic, done in 2 places, is that correct?

**Review by pierrejeambrun — APPROVED** [review]

**Review by hussein-awala — APPROVED** [review]
As a new contributor, you should read and follow [this doc](https://github.com/apache/airflow/blob/main/CONTRIBUTORS_QUICK_START.rst).

I'll merge the PR once the CI is passed.

**Comment by hussein-awala** [comment]
@potiuk, do you have any idea why some of the CI checks fail without being executed? Is it related to the changes made yesterday, or is it just a temporary problem?

**Comment by potiuk** [comment]
Likely intermittent. Looks like temporary Github Issue. Easiest is to rebase it or close/reopen such PR.

**Comment by hussein-awala** [comment]
It still has the same problem after rebasing and closing/reopening it.
I wonder if it's something related to a maintainer commit in no-maintainer PR :thinking:

**Comment by hussein-awala** [comment]
@potiuk I think it's indeed related, the job has `self-hosted` label, but it was blocked by the security policy. wdyt?
Here is the debugging log for one of the failed jobs:
```
Requested labels: self-hosted
Job defined at: apache/airflow/.github/workflows/ci.yml@refs/pull/31421/merge
Waiting for a runner to pick up this job...
Job is about to start running on the runner: Airflow Runner 62 (organization)
Error: Running job on worker Airflow Runner 62 disallowed by security policy
##[debug]Finishing: Push Early Image Cache (linux/amd64)
```

File: airflow/api_connexion/schemas/task_instance_schema.py
Comment (uranusjr) [inline airflow/api_connexion/schemas/task_instance_schema.py]: Let’s change these to use TaskInstanceState instead (also the one below)

Comment (JoranDox) [reply airflow/api_connexion/schemas/task_instance_schema.py]: > Let’s change these to use TaskInstanceState instead (also the one below)

Good point, I didn't dare but you're right, just did it.

**Review by uranusjr — APPROVED** [review]

**Comment by pierrejeambrun** [comment]
@JoranDox The PR is almost ready, can you fix the CI so we can merge it ?

**Comment by JoranDox** [comment]
> @JoranDox The PR is almost ready, can you fix the CI so we can merge it ?

Apologies, I thought the CI had issues unrelated to my code and was kind of ignoring the mailthread, thanks for pinging me! I forgot to import TaskInstanceState, but that should be all I think.

**Comment by JoranDox** [comment]
Why does black complain about files I didn't even edit? I can't even find them...
image

**Comment by JoranDox** [comment]
No idea what those sql failing tests have to do with my PR..

e.g. the fact that these two strings are not the same doesn't seem quite relevant to allowing the "skipped" state?

E             - http://localhost:8080/ javascript:alert(1)
E             + /home

**Comment by pierrejeambrun** [comment]
Make sure to rebase your branch, you are 25 commits behind, there's been work on those, and fix might already be in main.

**Pull request merged**
